### PR TITLE
Support RN 0.71.0

### PR DESCRIPTION
**Pull request by johnf**
React native now uses react-native-gradle-plugin, which works a bit differently.

We may need to support the old method as well

https://github.com/microsoft/react-native-code-push/issues/2418

### CODE REVIEW

File: android/codepush.gradle
Comment (johnf) [inline android/codepush.gradle:29]: Not sure if this is correct or if this is where it's supposed to fall into the other side of the if

Comment (DmitriyKirakosyan) [reply android/codepush.gradle:29]: @johnf , Does this condition fix something? Do we need it?

Comment (johnf) [reply android/codepush.gradle:29]: So the new React Native Gradle Plugin doesn't create a bundle task for debug builds.
So this was failing for me without it.
But this was the first change I made before I fixed the production builds.
It's possible that the else that's complimentary to this code is what's supposed to run in dev.
I'll see if I can do some more testing tomorrow and work out if this stays or now._

Comment (DmitriyKirakosyan) [reply android/codepush.gradle:29]: Got it. What about the back compatibility? Should we fallback to old code for older gradle plugin?

Comment (johnf) [reply android/codepush.gradle:29]: @DmitriyKirakosyan I'm about to push some change

Re the debug skip - the new plugin skips creating a bundle in debuggable builds entirely. Which makes sense. I think normally, you would be connecting to metro. I have used a similar code to the new plugin to skip code push in the same way.

The one thing I'm confused about is this code path https://github.com/microsoft/react-native-code-push/blob/master/android/codepush.gradle#L69-L108

For it to execute, ```bundleJsAndAssets``` has not to exist, but then in the first line after this else block, we depend on ```bundle${target}JsAndAssets```. So I can't understand how it could have ever worked in its current form. As far as I know, rails would always have included this task.

Is it historical and just hasn't been relevant for a while? I would be tempted to remove it entirely, but I'm not familiar enough to know if that's a good idea.

Comment (johnf) [reply android/codepush.gradle:29]: > Got it. What about the back compatibility? Should we fallback to old code for older gradle plugin?

If you look at the new changes (I discovered that the react config wasn't being pulled in) they are pretty substantial. I think maintaining backward compatibility would be difficult.

Given the 0.71 upgraders suggest switching everything across to the new plugin, I think we avoid backward compatibility and update the table in the Readme (which I've done)

Comment (Titozzz) [reply android/codepush.gradle]: I've seen other libs drop support for version below 0.71 due to the complexity of maintaining both so it seems reasonable to me

**Comment by Rag0n** [comment]
Do you have these warnings when building android?
```
Bugsnag: unable to find ReactNative bundle task 'bundleReleaseJsAndAssets'
Bugsnag: unable to find ReactNative bundle task 'bundleInternalReleaseJsAndAssets'
```

**Review by DmitriyKirakosyan — APPROVED** [review]

**Comment by gvillenave** [comment]
Can we get this merged?

**Review by Priskorn — APPROVED** [review]

**Comment by MrLibya** [comment]
Will this be merged anytime soon ?

**Comment by nikilarigela** [comment]
Please do it soon

**Comment by DisciplineBK** [comment]
when can get this merged?  Will it be too late?

**Comment by memon07** [comment]
can you please merge this

**Comment by FarukHossain2133** [comment]
Also i am facing this same issue after upgrading my react native project in "0.71.3"

**Comment by TM-TerryDutton** [comment]
I'm also facing this issue following an upgrade to react-native 0.71.3.

**Comment by johnf** [comment]
@DmitriyKirakosyan @Priskorn Anything else you need from me to get this merged and roll a release?

File: android/codepush.gradle
Comment (thewulf7) [inline android/codepush.gradle:6]: * What went wrong:
A problem occurred evaluating script.
> No signature of method: java.util.ArrayList.get() is applicable for argument types: () values: []
  Possible solutions: get(int), get(int), set(int, java.lang.Object), set(int, java.lang.Object), grep(), grep()

**Comment by thewulf7** [comment]
Also error in pipeline - compileSdkVersion is not specified. Please add it to build.gradle

File: android/codepush.gradle
Comment (thewulf7) [reply android/codepush.gradle:6]: you're mixing Kotlin into gradle with .get()

**Review by thewulf7 — CHANGES_REQUESTED** [review]

File: android/codepush.gradle
Comment (thewulf7) [inline android/codepush.gradle:71]: looks like different logic no?

**Comment by thewulf7** [comment]
Can't agree on dropping backward compatibility

**Comment by rufat** [comment]
For those having this issue and can't wait for merge, I have forked and published a variant of this library: `npm install react-native-code-push-next`
It supports 0.70.1 and above versions.

**Review by thewulf7 — APPROVED** [review]

**Comment by thewulf7** [comment]
> @DmitriyKirakosyan @Priskorn Anything else you need from me to get this merged and roll a release?

bump version in package.json

**Comment by sesm** [comment]
Can't wait for this to get merged!
Meanwhile anyone interested can test the fix this way:
```    
"react-native-code-push": "https://github.com/johnf/react-native-code-push#rn-0.71",
```

**Comment by johnf** [comment]
@thewulf7 
> Can't agree on dropping backward compatibility

On my quick exploration, I couldn't figure out a sensible way to do it. I'm also working on the assumption that most folks follow the reacgt-native upgrade guide and so they'll move to RNGP. If that's not the case we can always add it back later?

Or do you have access to any internal stats? How many react projects on 0.71 are using code push? If it's circa 50 I'd say they are using this patch, if it's more than that then perhaps they have upgraded and are using the old way

**Comment by sesm** [comment]
What are those test setup issues? Can we see them on other branches created from latest `master`? Maybe they are transient and just re-running CI tasks will help?
(sorry for jumping in, but I need this fix to complete my RN upgrade 😅 )

**Comment by MikhailSuendukov** [comment]
Hello @johnf, I managed to fix some of the tests, but after your changes, some tests are still failing. Could you please make sure that all tests pass before finalizing this pull request?

**Comment by johnf** [comment]
@MikhailSuendukov I think I've fixed the tests, they now work locally at least. (The test logic to include codeoush.gradle wasn't working with newer RN).

Can you approve so the tests run and then I'll rebase to clean up the commits.

I'm not sure those path changes were needed, I don't need them locally, but something might run differently in GitHub Actions. I have a debug print in there to help debug if it fails.

File: README.md
Comment (mk0116) [inline README.md:78]: Isn't it supposed to be 0.65-v0.70 ?

Comment (johnf) [reply README.md:78]: Yes that makes sense. I was fixing that line because the Code push version was wrong there is no 7.2, latest version is 7.1

File: android/codepush.gradle
Comment (MikhailSuendukov) [inline android/codepush.gradle:73]: It looks like [there](https://github.com/microsoft/react-native-code-push/pull/2419/checks#step:16:432) is a build error due to duplicate assets

**Review by ismailharmanda — APPROVED** [review]
Good job! It works. Please merge.

**Comment by DisciplineBK** [comment]
I hope it will be released soon.

File: android/codepush.gradle
Comment (Protino) [reply android/codepush.gradle:73]: @johnf Just a thought, would it be possible to run GitHub actions on your personal repo to verify what's wrong?

Comment (johnf) [reply android/codepush.gradle:73]: @Protino Thanks for reminding me of the obvious!! :smile: trying that now

**Comment by doyunnn** [comment]
Please merge asap!🥹

**Comment by stefan-schweiger** [comment]
@johnf @MikhailSuendukov seems like the android tests have failed :(

**Comment by MikhailSuendukov** [comment]
@johnf It seems that after your fixes, the tests stopped running at all, so I rolled back the changes inside the CLI file. As for the successfully passed tests in your repository: apparently, the tests stage turned out to be completely empty, so this stage was successfully completed. There are also problems with iOS tests.

File: android/codepush.gradle
Comment (johnf) [reply android/codepush.gradle:73]: @sesm were you switching between old and new versions? The old react native grade code and the new plugin seem to use different directories for flavours. debugUat vs debug/uat
So a clean is likely required in that instance

Comment (sesm) [reply android/codepush.gradle:73]: @johnf I've been switching between old and new versions, but I've done `./gradlew clean` since then, so all artefacts should be cleaned. I'm having an issue right now, that after I create a build with release configuration (`./gradlew assemblePlaygroundRelease` in my case`), I have to clean the project before I can create the same build one more time, otherwise I get a `duplicate assets` error.

**Comment by DisciplineBK** [comment]
This branch does not work with ios. right?

**Comment by Developeranees** [comment]
Please merge ASAP

**Comment by Developeranees** [comment]
I think now its very critical bcz expo sdk 48 got released and alot of people upgrading to latest sdk

**Comment by stefan-schweiger** [comment]
@Developeranees I agree that this is a pain point, but john is just a volunteer who took this upon him and if he doesn't find the time then he doesn't, simple as that. Microsoft on the other hand should maybe be a bit more interested in getting this done since they offer this as part of their product...

**Comment by b3itgustaf** [comment]
You're an absolute legend for working on this @johnf, thank you! <3 
I hope the microsoft-team can get involved soon to lend a helping hand and/or review this

File: android/codepush.gradle
Comment (nhatnguyenbp) [reply android/codepush.gradle:73]: I can resolve the issue `duplicate assets` by add these line to function `addJsBundleAndMetaToManifest` in `react-native-code-push/scripts/generateBundledResourcesHash.js` (under var savedResourcesManifestPath). The issue happen when start building project second time and we should remove `android/app/build/ASSETS/createBundleDebugJsAndAssets/CodePushHash` to make it works again

`             var savedResourcesManifestPath = assetsDir + "/" + CODE_PUSH_HASH_FILE_NAME;
+            if(fileExists(savedResourcesManifestPath)) {
+                fs.unlinkSync(savedResourcesManifestPath);
+            }`

**Comment by stefan-schweiger** [comment]
@Developeranees in case you are still struggling with this and expo 48 here is my current workaround until the PR gets merged https://github.com/microsoft/react-native-code-push/issues/2418#issuecomment-1469598851

**Comment by adamsinnott** [comment]
Is it possible to give some timeline as to when this might be merged in?

**Comment by thespacemanatee** [comment]
@thewulf7 you can't agree on dropping backward compatibility so you're just going to not support the latest version anyway? Help me understand the rationale, please.

**Comment by jkepps** [comment]
@johnf (or anyone else for that matter) are you experiencing issues on iOS? we are consistently having problems with the updates getting applied on iOS devices as well and i noticed this PR only changes android files

**Comment by stefan-schweiger** [comment]
@jkepps upgrading from expo 47 to expo 48 (so updating from react-native 0.69 to 0.71) did not have any effect on the behavior of CodePush for me. If you encounter any specific issues I would suggest to create a minimal reproduction and test with the last officially supported react-native version and the most recent one if there are really any 0.71 related issues.

**Review by ismailharmanda — DISMISSED** [review]

**Comment by johnf** [comment]
@jkepps Only android files are changed as the new release broke android due to changes in the way react native builds using gradle. I'm not aware of any related issues with iOS

**Comment by johnf** [comment]
@DmitriyKirakosyan @Priskorn  @thewulf7 

I've taken this as far as I can. I'm using this in production on Android and iOS without issues.

As far as the tests go
* Android - one broken test in CI, but it works fine when I run it locally.
* iOS - all the tests that create the code push with metro fail. I'm by no means an iOS developer and can't work out how to get the right logs locally on the device to see what's broken

I, unfortunately, can't spend any more time on this. It will need someone from Microsoft to the community to take it the rest of the way.

**Comment by akshgods** [comment]
> For those having this issue and can't wait for merge, I have forked and published a variant of this library: `npm install react-native-code-push-next` It supports 0.70.1 and above versions.

what is difference than regular one?
do we need to remove older one

**Comment by chj-damon** [comment]
finally!!

**Comment by gabrielbuzziv** [comment]
Using RN 0.64.4 and gettings the same error :/
`react-native-code-push` `8.0.1`

**Comment by chj-damon** [comment]
8.0.1 only support react-native v0.71 above.

File: android/codepush.gradle
Comment (cbargren) [reply android/codepush.gradle]: Am I understanding this right that v8.0.0+ will only support RN v0.71+ on Android? Right now the compatibility table in the README doesn't seem to indicate that if so.

**Comment by PratyakshGupta48** [comment]
After struggling with this error for 2 days , here is my final takeaway - 
1. Either you downgrade your react-native-codepush version to 7.1.0 or below if using react native version less than 0.71
2. Or you upgrade your react native version to 0.71 and above if you are using react-native-codepush version 8.0.1 and above.

**Pull request merged**
